### PR TITLE
[proposal] ospf layer

### PR DESCRIPTION
**Pull request by florianl**
Hi,
I've implemented the ospf layer for gopacket and I'm looking forward to answer your comments.

Kind regards,
 Florian

### CODE REVIEW

File: layers/ospf.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+package layers
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/gopacket"
+)
+
+type OSPFType uint8
```
Comment (gconnell) [inline layers/ospf.go]: Please add comments for all new types, and for exported functions (even if they are simply "DecodeFromBytes implements gopacket.DecodingLayer" or the like).

File: layers/ospf.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+// OSPFv3 extend the OSPF head with version 3 specific fields
+type OSPFv3 struct {
+	Instance uint8
+	Reserved uint8
+}
+
+// OSPF is a basic OSPF packet header with common fields of Version 2 and Version 3.
+type OSPF struct {
+	BaseLayer
+	OSPFv2
```
Comment (gconnell) [inline layers/ospf.go]: Interesting use of embedded types here.  I'm not sure if I prefer your approach:

```go
type OSPF struct {
  OSPFv2
  OSPFv3
  ...
}
```

or this approach:

```go
type OSPF {
  ...  common fields ...
}

type OSPFv2 {
  OSPF
  ... v2 specific fields
}

type OSPFv3 {
  OSPF
  ... v3 specific fields ...
}
```

I'd be interested to hear your thoughts on each.  A question this brings up is whether an OSPFv2 packet should be a different layer type than an OSPFv3 one.

File: layers/ospf.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	return gopacket.LayerTypePayload
+}
+
+// CanDecode returns the set of layer types that this DecodingLayer can decode.
+func (ospf *OSPF) CanDecode() gopacket.LayerClass {
+	return LayerTypeOSPF
+}
+
+func decodeOSPF(data []byte, p gopacket.PacketBuilder) error {
+
```
Comment (gconnell) [inline layers/ospf.go]: spurious empty line ;)

**Comment by florianl** [comment]
I signed it!

File: layers/ospf.go (hunk 2 above)
Comment (florianl) [inline layers/ospf.go]: My approach might cost a few more bytes in memory, but keeps it as simple as possible for the user. You just have to handle with one struct.

**Comment by florianl** [comment]
I've changed the code according the advice of @gconnell.

**Comment by florianl** [comment]
Hi, 
I've extended the PR, so not only the header of the OSPF packet is handled but also the whole packet as it is described in RFC 5340. So I'm looking forward your comments.

**Comment by florianl** [comment]
If there are problems or questions, please let me know @gconnell .

**Comment by florianl** [comment]
I've resolved conflicts because of recent changes.

**Comment by gconnell** [comment]
Could you merge layers/layertypes and resubmit?

**Comment by gconnell** [comment]
Thanks for your great work on this!

**Pull request merged**